contracts/AssetPool.sol

```diff
@@ -80,17 +124,112 @@ contract AssetPool is Ownable {
+        //
+        //      withdrawal              graceful withdrawal
+        //        delay                 timeout
+        //  /--------------\ /----------|
+        // x----------------x-----------x-------------------------------x------>
+        // ^                 \-----------------------------------------|
+        // initiatedAt                                   hard withdrawal
+        //                                               timeout
+        //
+
+        // When the graceful withdrawal time ends. After this time, part of the
+        // collateral and rewards will be seized by the pool.
+        uint256 gracefulWithdrawalEndTimestamp =
+            withdrawalDelayEndTimestamp.add(gracefulWitdrawalTimeout);
+        // When the time for the withdrawal ends. After this time, 99% of
+        // rewards and collateral is seized by the pool and 1% of rewards and
+        // collateral is sent to the notifier who completed the withdrawal on
+        // behalf of the underwriter.
+        uint256 hardWithdrawalEndTimestamp =
+            withdrawalDelayEndTimestamp.add(hardWithdrawalTimeout);
+
+        if (gracefulWithdrawalEndTimestamp >= block.timestamp) {
+            // Before the graceful withdrawal timeout. This is the happy path.
+            emit WithdrawalCompleted(
+                underwriter,
+                amountToWithdraw,
+                block.timestamp
+            );
+            collateralToken.safeTransfer(underwriter, amountToWithdraw);
+        } else if (hardWithdrawalEndTimestamp >= block.timestamp) {
+            // After the graceful withdrawal timeout but before the hard
+            // withdrawal timeout. A portion of collateral and tokens is
+            // seized by the pool, proportionally to the time passed after
+            // the graceful withdrawal timeout.
+            uint256 delayRatio =
+                hardWithdrawalEndTimestamp.sub(block.timestamp).mul(1e18).div(
+                    hardWithdrawalTimeout.sub(gracefulWitdrawalTimeout)
+                );
+            // slither-disable-next-line divide-before-multiply
+            uint256 amountToWithdrawReduced =
+                delayRatio.mul(amountToWithdraw).div(1e18);
+            emit WithdrawalCompleted(
+                underwriter,
+                amountToWithdrawReduced,
+                block.timestamp
+            );
+            emit GracefulWithdrawalTimedOut(underwriter, block.timestamp);
+            collateralToken.safeTransfer(underwriter, amountToWithdrawReduced);
+        } else {
+            // After the hard withdrawal timeout passed. 99% of tokens is seized
+            // by the pool, 1% of tokens goes go the notifier.
+            emit WithdrawalCompleted(underwriter, 0, block.timestamp);
+            emit WithdrawalTimedOut(underwriter, block.timestamp);
+            collateralToken.safeTransfer(msg.sender, amountToWithdraw.div(100));
```

Comment:
`amountToWithdraw.div(100)` I take that we operate on 1e18 numbers here and losing precision here is ok?

Reply:
In this particular case, I do not think there is much we can improve. We could, theoretically, improve the precision of the previous case by computing `amountToWithdraw` inside the `else if` and not dividing by `covSupply` before calculating `delayRatio` and I tried to refactor the code this way but... it is getting really, really clumsy to understand and two other detectors start complaining that the code looks very similar in if/else blocks... Given that the vast majority (if not all) underwriters should complete the withdrawal in the first 7 days, I decided to pick the clarify of the code at the expense of precision. 🤷